File: Python/hashmap.py
# Define the HashMap class
class HashMap:
    # Constructor method to initialize the hash map
    def __init__(self):
        self.size = 1000  # Set the size of the hash table to 1000
        self.table = [[] for _ in range(self.size)]  # Create an array of empty lists (buckets) for chaining

    # Private method to compute hash index for a given key
    def _hash(self, key):
        return key % self.size  # Use modulo operation to find the bucket index

    # Method to insert or update a key-value pair in the hash map
    def put(self, key, value):
        idx = self._hash(key)  # Get the bucket index for the key
        for pair in self.table[idx]:  # Iterate through the bucket
            if pair[0] == key:  # If key already exists
                pair[1] = value  # Update the value
                return  # Exit after updating
        self.table[idx].append([key, value])  # If key not found, append new key-value pair to the bucket

    # Method to retrieve the value for a given key
    def get(self, key):
        idx = self._hash(key)  # Get the bucket index for the key
        for pair in self.table[idx]:  # Iterate through the bucket
            if pair[0] == key:  # If key matches
                return pair[1]  # Return the value
        return -1  # If key not found, return -1

    # Method to remove a key-value pair from the hash map
    def remove(self, key):
        idx = self._hash(key)  # Get the bucket index for the key
        for i, pair in enumerate(self.table[idx]):  # Iterate with index through the bucket
            if pair[0] == key:  # If key matches
                self.table[idx].pop(i)  # Remove the pair from the bucket
                return  # Exit after removal


# Define the main function to demonstrate the HashMap usage
def main():
    print("HashMap Demo")  # Print title
    hm = HashMap()  # Create a new HashMap instance
    hm.put("apple", 5)  # Insert key 'apple' with value 5
    hm.put("banana", 7)  # Insert key 'banana' with value 7
    print("Get apple:", hm.get("apple"))       # Should print 5
    print("Get banana:", hm.get("banana"))     # Should print 7
    print("Get orange:", hm.get("orange"))     # Should print -1 (not found)
    hm.remove("apple")  # Remove key 'apple' from the map
    print("Get apple after removal:", hm.get("apple"))  # Should print -1 (after removal)

# Execute the main function if this script is run directly
if __name__ == "__main__":
    main()

File: JavaScript/hashmap.js
// JavaScript: HashMap class
class MyHashMap {
    constructor() {
        // Set the size of the bucket array
        this.size = 1000;
        // Create an array of empty arrays (buckets) to store key-value pairs
        this.buckets = Array(this.size).fill(null).map(() => []);
    }

    // Hash function: returns an index for a given key
    hash(key) {
        return key % this.size;
    }

    // Inserts or updates a key-value pair
    put(key, value) {
        // Get the correct bucket based on the hash of the key
        let bucket = this.buckets[this.hash(key)];

        // Loop through the bucket to see if the key already exists
        for (let pair of bucket) {
            if (pair[0] === key) {
                // If key exists, update the value
                pair[1] = value;
                return;
            }
        }

        // If key doesn't exist, push the new key-value pair into the bucket
        bucket.push([key, value]);
    }

    // Retrieves the value for a given key
    get(key) {
        // Get the correct bucket based on the hash of the key
        let bucket = this.buckets[this.hash(key)];

        // Search the bucket for the key
        for (let pair of bucket) {
            if (pair[0] === key) {
                // Return the value if key is found
                return pair[1];
            }
        }

        // If key is not found, return -1
        return -1;
    }

    // Removes the key-value pair for a given key
    remove(key) {
        // Get the correct bucket based on the hash of the key
        let bucket = this.buckets[this.hash(key)];

        // Filter out the pair with the matching key
        this.buckets[this.hash(key)] = bucket.filter(pair => pair[0] !== key);
    }
}

// ---------------- Main ----------------

// Create a new instance of MyHashMap
const map = new MyHashMap();

// Insert key-value pair (1, 10)
map.put(1, 10);

// Insert key-value pair (2, 20)
map.put(2, 20);

// Insert key-value pair (102, 999)
// This may collide with key 2 if the hash function results in the same index
map.put(102, 999);

// Retrieve value for key 1 and print it (should be 10)
console.log(map.get(1));     // Output: 10

// Retrieve value for key 2 and print it (should be 20)
console.log(map.get(2));     // Output: 20

// Retrieve value for key 102 and print it (should be 999)
console.log(map.get(102));   // Output: 999

// Remove key 2 from the map
map.remove(2);

// Try to retrieve value for key 2 again (should be -1 since it was removed)
console.log(map.get(2));     // Output: -1
